test/workbox-precaching/sw/test-createHandlerForURL.mjs

```diff
@@ -0,0 +1,77 @@
+/*
+  Copyright 2019 Google LLC
+
+  Use of this source code is governed by an MIT-style
+  license that can be found in the LICENSE file or at
+  https://opensource.org/licenses/MIT.
+*/
+
+import {createHandlerForURL} from 'workbox-precaching/createHandlerForURL.mjs';
+import {precache} from 'workbox-precaching/precache.mjs';
+
+describe(`createHandlerForURL()`, function() {
+  const sandbox = sinon.createSandbox();
+
+  beforeEach(function() {
+    sandbox.stub(self, 'addEventListener');
+  });
+
+  afterEach(function() {
+    sandbox.restore();
+  });
+
+  it(`should throw when passed a URL that isn't precached`, function() {
+    precache([]);
+
+    return expectError(() => {
+      createHandlerForURL('/does-not-exist');
+    }, 'non-precached-url', (error) => expect(error.details.url).to.eql('/does-not-exist'));
+  });
+
+  it(`should return the expected handlerCallback for precached URLs`, async function() {
+    // Simulate the following: first two handlerCallbacks have cache.match()
+    // calls that return a hit. Third handlerCallback has a cache.match() call
+    // that is a miss, which will lead to a call to fetch().
+    const matchStub = sandbox.stub()
+        .onFirstCall().resolves(new Response('response 1'))
+        .onSecondCall().resolves(new Response('response 2'))
+        .onThirdCall().resolves(undefined);
+
+    sandbox.stub(self.caches, 'open').resolves({
+      match: matchStub,
+    });
+
+    const fetchStub = sandbox.stub(self, 'fetch').resolves(new Response('response 3'));
+
+    precache([
+      '/url1',
+      {url: '/url2', revision: 'abc123'},
+      '/url3',
+    ]);
+
+    const handler1 = createHandlerForURL('/url1');
+    const response1 = await handler1();
+
+    expect(matchStub.calledOnce).to.be.true;
+    expect(matchStub.firstCall.args).to.eql([`${location.origin}/url1`]);
+    expect(fetchStub.notCalled).to.be.true;
+    expect(await response1.text()).to.eql('response 1');
+
+    const handler2 = createHandlerForURL('/url2');
+    const response2 = await handler2();
+
+    expect(matchStub.calledTwice).to.be.true;
+    expect(matchStub.secondCall.args).to.eql([`${location.origin}/url2?__WB_REVISION__=abc123`]);
+    expect(fetchStub.notCalled).to.be.true;
+    expect(await response2.text()).to.eql('response 2');
+
+    const handler3 = createHandlerForURL('/url3');
+    const response3 = await handler3();
+
+    expect(matchStub.calledThrice).to.be.true;
+    expect(matchStub.thirdCall.args).to.eql([`${location.origin}/url3`]);
+    expect(fetchStub.calledOnce).to.be.true;
+    expect(fetchStub.firstCall.args).to.eql([`${location.origin}/url3`]);
```

Comment:
Similar to my question above, do you also want to test the case with the fetch fallback for the cache key (assuming that's correct)?

Reply:
I can add that, sure.